### PR TITLE
Test result icon for test groups

### PR DESCRIPTION
**Pull request by rowo360**
I propose a solution in this PR for issue #1118. The current state is still not the final one, because all the tests are missing yet (and I expect that these scenarios are good testable). But let's discuss the proposed solution first.

I split the solution into three separate commits, hoping that it's easier to keep an overview.
1. commit: Reactivate existing solution to update tree icons for category groups
2. commit: Support update of tree icons for outcome and duraction groups
3. commit: Enhance update of tree icons for category nodes

I'll add some details in the next comments.

### CODE REVIEW

**Comment by rowo360** [comment]
Some details about the **first commit**, which has the focus to update tree icons for **category groups**.

The essential part is in the class `GroupDisplayStrategy` - now the method `_grouping?.OnTestFinished(result)` is called instead of the method `ApplyResultToGroup(result, false)`. Unfortunately it's not visible in the diff-view, but we need to look to the code ourselves: The three grouping classes (CategoryGrouping, OutcomeGrouping and DuractionGrouping) already implemented the method `OnTestFinished`, but that code was not invoked anymore so far.

Looking into the git history I spotted commit #865 which was dealing with the elimination of the TreeViewElement and ITreeView. This commit exactly removed the call to `_grouping?.OnTestFinished(result)` and I'm rolling it back, now. But I think, it's for good. :-)

One concern might be that the method `ApplyResultToGroup(result, false)` is not called anymore for category grouping. But I think that it's for the best. Well, I already noticed some strange behavior sometime ago and wanted to raise a separate issue about it at times. Now I happen to stumble across the exact code that is responsible for this. Here's a screenshot about it:



I started a test run for one category (DisplayFormat: Testlist, GroupBy: Category (extended)). After the test run the category 'Feature_2' is not displayed anymore in tree and the test cases are listed under category 'None'. This behavior might be related to our discussion in #1106 , but actually the method `ApplyResultToGroup(result, false)` is responsible for moving nodes between group nodes. 
That's of course necessary for Duration and OutCome grouping, but the test categories will never change dynamically during a test run. That's why I think it's really right to stop using this method for categories - and it also fixes this strange behavior!

Next, I learned that there's an ImageIndex also in the TestGroup class. Sometime ago we decided to reset all ImageIndex at the start of a test run, but I missed to reset these ImageIndex properties in the TestGroups. And now these outdated ImageIndex at the TestGroups do some harm. Therefore I introduced a OnTestRunStarting() method at the Strategy class, which forward the method to the TestGrouping, which can perform the reset.

With these changes the existing method `CategoryGrouping.OnTestFinished()` gets invoked again, which already implemented the update of the tree icons for category nodes. I am almost satisfied with the existing behavior. The icon is always the right one! And I think this is the most important thing.
Less important, but still noticeable, is that the icon is set a little too early. Currently it's already set when the first test fixture (Fixture list mode) or first test case (Test list mode) is finished. It might be updated when the resulting state changes with some subsequent tests. 

But overall I expect that the icon should only be updated when the last test in this category has been completed. I think that's how I know the behavior so far. But let's discuss - I tried an improved solution with commit no.3

**Comment by rowo360** [comment]
Some details about the **second** commit, which has the focus to update tree icons for **outcome** and **duration** groups.

Actually, the solution for the categories also works for outcome and duration grouping. So one solution might be to move that code from CategoryGrouping to GroupingStrategy class, so that it can be used for all three kind of groupings. I made an attempt and unfortunately notice now, looking at the commit, that I missed to remove that code again (method `UpdateTestResult(ResultNode result)` - it's dead code, now).
But the little drawback of this solution is also noticable here: for example the icon of node 'Passed' gets immediately updated after the first test finished. Therefore I made an attempt for a different solution:

It's based on the consideration that I can only actually set the icon correctly once the last test has been run. I cannot predict if a test will run 'slow' or 'fast' or if a test will 'pass' or 'fail' and finally in which group that test belongs to. Therefore, it is impossible to determine the correct time when the last test of one of these groups was executed and you can now set the icon for this group. 

Therefore I suggest to set the icon of these group nodes at the end of a test run!
Technically I'm handling the RunFinished event, passing it from the TreeViewPresenter via the Strategy to the Grouping. And most important the Grouping class can use an existing method `Strategy.CalcImageIndexForGroup(group)` for determining the group icon. So this essential logic must not be implemented.

**Comment by rowo360** [comment]
Some details about the **third** commit, which enhance the update of tree icons for **category groups**.

I think that the icon of category nodes can either be updated as soon as a test result for all children is present or at the end of the test run. Maybe I'm oversimplifying it. And it shouldn't be 'test result for all children', but 'last test result' of that category. (imagine that I execute only a subset of tests of a category). But I have the feeling that it will be difficult to determine that a test result is really the last one of one category.

Therefore I stick with the 'test result for all children' approach and I consider it as an improvement compared to setting the icon too early.

**Comment by rowo360** [comment]
Sorry for the long comments and all the text - I hope it helps with understanding and discussion.

**Comment by CharliePoole** [comment]
This will take me a while to review. :-)

File: src/TestCentric/testcentric.gui/Presenters/CategoryGrouping.cs [hunk 1]
```diff
@@ -52,13 +52,17 @@ public override void OnTestFinished(ResultNode result)
...
-                            parentNode.SelectedImageIndex = parentNode.ImageIndex = group.ImageIndex = imageIndex;
-                        }
+                        continue;
+                    }
+
+                    // Category nodes are updated either as soon as a test result for all children is present or at the end of the test run (method OnTestRunFinished)
+                    if (TestResultExistsForAllChildren(group))
+                    {
+                        int groupImageIndex = _displayStrategy.CalcImageIndexForGroup(group);
+                        parentNode.SelectedImageIndex = parentNode.ImageIndex = group.ImageIndex = imageIndex;
```
Comment (CharliePoole) [inline src/TestCentric/testcentric.gui/Presenters/CategoryGrouping.cs:65]: Throwing out an idea here... tell me what you think.

What if all the updating of parent nodes was handled by the view? Of course, we use the presenter to make testing easier but it adds complexity too. My logic is as follows:
1. Everything in the tree has an index and that index is supposed to show the "worst" result from all its children. It seems like it should not matter whether the parent is another test, a category or any other group node.
2. So what  if we did this update to higher levels at the view level, dealing only with tree nodes?
3. When a node's index is being updated, that node could propagate the same index to all higher level nodes IFF the level of the node were not already worse than that update.
4. We would work our way up the tree until there was no parent OR until no update was needed.
 
Would this be simpler than what we do now?

File: src/TestCentric/testcentric.gui/Presenters/CategoryGrouping.cs [hunk 2]
```diff
@@ -113,6 +117,26 @@ private void AddGroup(TestGroup group)
...
+                var imageIndexInGroup = DisplayStrategy.CalcImageIndex(result.Outcome);
+                if (imageIndexInGroup < TestTreeView.SuccessIndex)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
```
Comment (CharliePoole) [inline src/TestCentric/testcentric.gui/Presenters/CategoryGrouping.cs:139]: Seems like a lot of work per update, but see earlier comment about letting the view do it.

File: src/TestCentric/testcentric.gui/Presenters/DisplayStrategy.cs [hunk 3]
```diff
@@ -83,6 +83,15 @@ public virtual void OnTestFinished(ResultNode result)
                 _view.SetImageIndex(treeNode, imageIndex);
         }
 
+        public virtual void OnTestRunStarting()
+        {
+            _view.ResetAllTreeNodeImages();
```
Comment (CharliePoole) [inline src/TestCentric/testcentric.gui/Presenters/DisplayStrategy.cs:88]: Is this needed for the NUnit tree?

**Review by CharliePoole — CHANGES_REQUESTED** [review]
Two main changes...

1. I can see we are getting a lot of complications in the code due to the introduction of multiple strategies and because we are trying to use MVP properly. In the original NUnit GUI (V2), we simply gave the parent node the "worst" color index comparing its present setting to the setting of a child test that had just completed. This depends on the numerical order of the indices and on the policy of always propagating errors up the tree display. 

You could either do this in the view or in the presenter itself, working only with tree nodes. Either approach is less "pure" than what we are doing now, but even so might be simpler. 

If this makes sense to you, I suggest trying this change before the following suggestion because I suspect it will simplify the remaining code.

2. As we discussed this AM, let's just have a Category grouping in the UI rather than Category and Category (extended). In CategoryGrouping, we should change the semantics so it works like Extended Category in the TestList strategy only. IOW, if the user asked for a TestList, we will assume they want Categories to work as if they are inherited. I don't think this will be surprising to most people.

File: src/TestCentric/testcentric.gui/Presenters/TreeViewPresenter.cs [hunk 4]
```diff
@@ -99,11 +99,16 @@ private void WireUpEvents()
...
                 _model.ClearResults();
-                _view.ResetAllTreeNodeImages(); 
+                Strategy.OnTestRunStarting();
                 CheckPropertiesDisplay();
                 CheckXmlDisplay();
             };
 
+            _model.Events.RunFinished += (ea) =>
+            {
+                Strategy?.OnTestRunFinished();
```
[the remark this answers is not in the document]
Comment (rowo360) [reply src/TestCentric/testcentric.gui/Presenters/TreeViewPresenter.cs:109]: Yes indeed, I can remove the '?' operator here. I was probably a little too careful - I'll remove it and double-check it.

File: src/TestCentric/testcentric.gui/Presenters/DisplayStrategy.cs (hunk 3 above)
Comment (rowo360) [reply src/TestCentric/testcentric.gui/Presenters/DisplayStrategy.cs:88]: Yes, this is also required for the NUnit tree. When starting a test run, we want that the images of all nodes are reset to the default (regardless of the current display format: NUnit, FixtureList or TestList). 
This statement was previously located in the TreeViewPresenter.WireUpEvents() - RunStarting event handler, but I moved it along to the new method Strategy.OnTestRunStarting().

File: src/TestCentric/testcentric.gui/Presenters/CategoryGrouping.cs (hunk 1 above)
Comment (rowo360) [reply src/TestCentric/testcentric.gui/Presenters/CategoryGrouping.cs:65]: I agree with you that this code part is somehow complex and we should consider how to simplify it.
If I look to the implementation of `TestGrouping.OnTestFinished`, it looks pretty close to your suggestion. That implementation is using the existing method `GroupDisplayStrategy.CalcImageIndexForGroup` - ok, that's not the 'bubble up' approach you mentioned, but instead an approach to get the 'worst' test result of any children.

But overall, I consider this approach used for the DurationGrouping and OutcomeGrouping as rather simple. We just need the OnTestRunFinished() event and update the tree node images of the 'Passed', 'Fast'... nodes at this point of time.

One idea: I could remove the entire `CategoryGrouping.OnTestFinished` method! That means that there's no difference between those Groupings anymore. I think that would simplify the code also! There's only one drawback from user point of view, that the image of a category node is updated only at the end of the test run and not immediately when the last test case of that category is finished.
That's not 100% accurate, but I believe it's acceptable. If users complain heavily about it, it can be improved at that point in time. Is this suggestion acceptable for you?

I have also thought about the 'bubble up' approach, which I expect to work properly. We only have to observe or decide at what time the parent should be updated? Is it right away when the first test case result of any child arrives? And of course it must be updated as soon as further test results arrive. That's different than the current behavior, but also fine!
Or we update the parent node only at that point of time, when the test case result of the last child arrives - but that starts to get complicated again.

Maybe I'm already too involved in the implementation, to get a clear picture.
But currently I propose just to remove `CategoryGrouping.OnTestFinished` method to reduce complexity.

Comment (CharliePoole) [reply src/TestCentric/testcentric.gui/Presenters/CategoryGrouping.cs:65]: Seems like a reasonable compromise. We could then live with it through the next beta and see how it seems both to us and users.

**Comment by rowo360** [comment]
The context menu 'By Category (Extended)' is removed now from user point of view.



Internally the category grouping still supports both modes (normal and extended mode). This was and is still controlled by a constructor parameter 'includeAncestors'. The only difference is that this parameter is set now depending on the current Strategy (FixtureList vs TestList).

I also checked the system behavior if a VisualState file with 'Category_Extended' mode is loaded. In that case the system will use the default grouping ('Assembly') - but most important the system doesn't crash.

I also removed the 'Category_Extended' handling from some tests. These tests didn't start to fail, but I prefer to clean up those right away.

**Review by CharliePoole — APPROVED** [review]
This is looking good to me. I'm going to bed and you're still working, so I'm approving now. When you're ready, merge if GitHub permits, which is what I hope I managed to get set up. If not, just remove the draft status and I'll do it in the morning.

**Comment by rowo360** [comment]
I added some tests for the different grouping and strategy classes. Some of them are pure unit tests (testing a single class) - some of them are testing a set of classes.

**Comment by rowo360** [comment]
Actually, I thought I was really done with this task now. But while testing I noticed an edge case, which I like to solve right away.
It's possible that a user defines the same test category at the test fixture **and** the test case!

Unfortunately, this means that the test cases are inserted twice when working with 'Test list' mode and 'Category' grouping.


This behavior is caused by the CategoryGrouping class which will find the category at the test case and the test fixture.

**Comment by CharliePoole** [comment]
I agree. Look at the code that creates the filter for the category. It should not be creating duplicates in the selection.
